rfc/rfc-0000-allow-resources-to-report-status.md

```diff
@@ -0,0 +1,298 @@
+    
+* If `alwaysHealthy: {}`, `reason: AlwaysHealthy`.
+* If `singleConditionType: X`, `reason: XCondition`.
+* If `multiMatch: ...`, `reason: (MatchedCondition|MatchedField)`.
+
+If the `healthRule` is not `alwaysHealthy`, the condition will also have a message. For `singleConditionType`, the message will be propagated from the specified condition. For `multiMatch`, the message will specify the value of the matched condition or field, and then display the message of the matched condition or the message at the path specified in `matchFields`.
+
+In the case of `multiMatch`, if there is more than one matching condition or field that caused the resource to enter it's current state, the first one read will be propagated by Cartographer.
+
+Each owner will display a new condition in the status called `ResourcesHealthy` which will only be updated to `True` when all resources in the blueprint are healthy. Cartographer will not update the top level state of an owner to `Ready: True` until `ResourcesHealthy` is `True`.
+
+# Migration
+[migration]: #migration
+
+Given that Cartographer will default to `Healthy` once the resource is applied and outputs are read, this will not be a breaking change.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Why should we *not* do this?
+- We only allow for reflecting one condition (`Healthy` as True/False/Unknown), though this design allows for addition conditions to the resources in the future.
+- This design does not allow authors to define the conditions themselves.
+  - This consistency is itself a positive, though, as the owner can reason about the state of the stamped resources.
+
+# Alternatives
+[alternatives]: #alternatives
+
+- What other designs have been considered?
+  - We could create a separate CRD to define the healthy condition rules and reference them from the templates or have them select templates based on the underlying resource type.
+    - Separating these rules in a separate CRD would force us to start reconciling templates to see if there is an associated healthy condition rule CR. 
+    - This would also force template authors to ship the healthy condition CR along with each of their templates.
+    - If CR selected templates, this could be a bit confusing because the type of the stamped resource may be templated as well. Therefore, a template could end up stamping out different types of resources and so Cartographer would only be able to inform users which healthy condition rule is used for a given template after an owner is applied, and it could then be reflected in the owner status.
+      - It would also mean that templates stamping out the same underlying type would not be able to define different healthy rules, if they desired to.
+- What is the impact of not doing this?
+  - Users/UI creators will have to have in depth knowledge of the underlying resources to determine the health of the owner.
+
+# Prior Art
+[prior-art]: #prior-art
+
+N/A
+
+# Unresolved Questions
```

Comment:
Does Cartographer need to consider observedGeneration?

Some prior art is the working of the `ClusterDeploymentTemplate`. In that resource, authors of the template define rules to specify when the object has completed work. We specify different manners to ensure that the status we are observing is for the spec that has been submitted. Should users be alerted if the spec and status are out of sync (e.g. spec is generation N while status reports generation N-1)? If they are out of sync in that manner, is the resource still `Healthy`?

Reply:
Good question. If we only care to surface the most recent health info we don't have to consider the observed generation. If we want to link that health to the spec, then we do.  Even if we don't care to link the health and specs, if our users want to link them then we should surface it.

Reply:
This is really the main responsibility of [another RFC](https://github.com/vmware-tanzu/cartographer/pull/799).
Resolving convo.